### PR TITLE
2FA: Refactor TwoStepAuthorization to redux

### PR DESCRIPTION
**Pull request by yurynix**
This PR refactors TwoStepAuthorization to redux

screen shot 2018-03-06 at 12 39 23

screen shot 2018-03-06 at 12 39 58

  
#### Testing instructions
* Use an account that has 2FA enabled
* For always triggering reauth, apply server side patch, or intercept call to `/me/two-step/` (i.e. pb-1d4e2 )

1. Run `git checkout update/reduxify-two-step` and start your server, or open a [live branch](https://calypso.live/?branch=update/reduxify-two-step)
2. Open the [`Me` page](http://calypso.localhost:3000/me)
3. Assert the 2FA dialog shows up when the page is loaded
4. Assert that entering correct Authenticator code dismisses the dialog
5. Assert that requesting the SMS sends you an SMS
6. Assert that entering the code you've got via SMS dismisses the dialog
  

#### Reviews
  
- [ ] Code
- [ ] Product
- [ ] Tests

### CODE REVIEW

**Comment by gwwar** [comment]
Thanks for taking this on! The diff looks rather large here already so I'd suggest splitting this up into a few PRs to make progress on this faster. Splitting things up makes it easier to review, and has much less risk to land each individual PR.
- [ ] A Janitorial PR to convert old code to new styles, such as not using React.createClass (This can also be done last, whatever is easier)
- [ ] Add new Redux reducers in parallel (but not wired in)
- [ ] A PR to switch update the interface to use Redux 
- [ ] A PR to remove any last old data pieces.

File: client/me/reauth-required/index.jsx [hunk 1]
```diff
@@ -254,17 +262,33 @@ const ReauthRequired = createReactClass( {
 				</form>
 			</Dialog>
 		);
-	},
+	}
 
-	handleChange( e ) {
+	handleChange = e => {
```
Comment (rodrigoi) [inline client/me/reauth-required/index.jsx]: you can do 
```js
handleChange = ( { currentTarget } ) => {}
```
or
```js
handleChange = ( { currentTarget: { name, value } } ) => {}
```
if you want to push syntax optimization even further. Completely optional.

File: client/me/reauth-required/index.jsx [hunk 2]
```diff
@@ -254,17 +262,33 @@ const ReauthRequired = createReactClass( {
...
 		const { name, value } = e.currentTarget;
 		this.setState( { [ name ]: value } );
-	},
+	};
 
-	handleCheckedChange( e ) {
+	handleCheckedChange = e => {
```
Comment (rodrigoi) [inline client/me/reauth-required/index.jsx]: same as https://github.com/Automattic/wp-calypso/pull/22984/files#r172963185

File: client/state/data-layer/wpcom/me/two-step/index.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+};
+
+const validateTwoStepCode = ( { dispatch }, action ) =>
+	dispatch(
+		http(
+			{
+				apiVersion: '1.1',
+				method: 'POST',
+				path: '/me/two-step/validate',
+				body: {
```
Comment (rodrigoi) [inline client/state/data-layer/wpcom/me/two-step/index.js]: if you destructure action into `{ code, action, remember2fa }` you can add a bit of `lodash` magic

```js
body: omitBy( { code, action, remember2fa }, isNil )
```

File: client/state/data-layer/wpcom/me/two-step/index.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+			},
+			action
+		)
+	);
+
+const storeTwoStepCodeValidationResult = ( { dispatch, getState }, action, data ) => {
+	// If the validation was successful AND reauth was required, fetch
+	// data from the following modules.
+	if ( data.success && isReauthRequired( getState() ) ) {
+		userSettings.fetchSettings();
```
Comment (rodrigoi) [inline client/state/data-layer/wpcom/me/two-step/index.js]: candidates for future _reduxification_ 😛

**Comment by rodrigoi** [comment]
if you do as @gwwar suggests, and make a PR for the actions/reducer/selectors and another for the data layer, you can add some unit tests 🙂

File: client/state/data-layer/wpcom/me/two-step/index.js (hunk 4 above)
Comment (yurynix) [reply client/state/data-layer/wpcom/me/two-step/index.js]: Yeah, would rather not touch it here =)

**Comment by yurynix** [comment]
Thank you for taking a look!

Regarding the split, it'll be 2 PRs:
- Redux + data layer
- wiring of the UI + migrating to ES6 class

Because:

> A Janitorial PR to convert old code to new styles, such as not using React.createClass (This can also be done last, whatever is easier)

This class using a mixin to get changes from the flux store `mixins: [ observe( 'twoStepAuthorization' ) ]`
to migrate from the mixin, we first need to move the functionlaity to redux, so can't touch that unless we have all the "backend" in-place.

> Add new Redux reducers in parallel (but not wired in)

I'll make a PR with data layer + redux not wired in, as data layer needs redux's actions to store retreived data and later follow up with a wiring PR.

**Comment by alisterscott** [comment]
This PR needs a rebase - is this PR still required?

**Comment by alisterscott** [comment]
Closing due to inactivity - please reopen if necessary

**Pull request closed without merging**